### DISCUSSION TITLE
How exactly do you use the pyi_splash module?

### DISCUSSION THREAD

**Discussion by aliencaocao** [pyinstaller/pyinstaller#6174] [Help] [answered]
I am able to add a splash screen image but it won't close at all, so I realised I have to use the pyi_splash module to control it, but after a very long time of experimentation, I still could not make any progess.

I am very confused by the documentation. It simply used import pyi_splash but python throws import error on the module. Then I saw I need to wrap it inside a try except, so I did, but with try except, how do I then refer to it in my program code and use the close() or update_text() methods, since it is not defined due to failed import? I have read the docs letter by letter and could not make any sense out of it. Can someone please tell me clearly how to actually customize this splash screen? Thanks!
The two docs I read: https://pyinstaller.readthedocs.io/en/latest/advanced-topics.html#pyi-splash-module
https://pyinstaller.readthedocs.io/en/stable/usage.html#splash-screen-experimental

(Sorry if I sounded very frustrated - because I have been stuck on this for very long and I saw no one asking anything similar to this question)

**Comment by rokm** [comment]
Hmm, well, one working example is in our test suite; [source](https://github.com/pyinstaller/pyinstaller/blob/develop/tests/functional/scripts/pyi_interact_pyi_splash.py) and corresponding [spec](https://github.com/pyinstaller/pyinstaller/blob/develop/tests/functional/specs/spec_with_splash.spec).

To turn it into (slightly changed) stand-alone example:
```python
# splash_example.py

from time import sleep


def main():
    # Init pyi_splash / connect to the bootloader.
    import pyi_splash

    # Modify text
    for i in reversed(range(5)):
        print(f"Updating text (1st time) in {i+1} seconds...")
        sleep(1)
    pyi_splash.update_text("This is a test text")

    # Modify text again
    for i in reversed(range(5)):
        print(f"Updating text (2nd time) in {i+1} seconds...")
        sleep(1)
    pyi_splash.update_text("Second time's a charm")

    # Close the splash screen
    for i in reversed(range(5)):
        print(f"Closing splash screen in {i+1} seconds...")
        sleep(1)
    pyi_splash.close()

    # Exit application
    for i in reversed(range(10)):
        print(f"Exiting application in {i+1} seconds...")
        sleep(1)


if __name__ == '__main__':
    main()
```

Generate the spec with custom splash image: `pyi-makespec splash_example.py --onedir --splash my-splash.jpg`

To make text appear, you need to modify its position in the spec file. I.e., change `text_pos=None,` to for example `text_pos=(20, 20),` (and to change color from black to red, you can add `text_color='red',`). 

Then build from the spec: `pyinstaller --clean --noconfirm splash_example.spec`

The frozen application should change text after five seconds (twice), close the splash screen after five seconds, and exit after another ten seconds.

**Reply by rokm** [reply]
Of course, this works only in *frozen* application. If you want your code to be still executable as source script under python, you need to put splash screen stuff into a [frozen-application-specific codepath](https://pyinstaller.readthedocs.io/en/stable/runtime-information.html?run-time-information#run-time-information), or interleave your general code with frozen-application-specific blocks, for example:

```python
# splash_example.py

import sys
from time import sleep

is_frozen = getattr(sys, 'frozen', False)

def main():
    # Init pyi_splash / connect to the bootloader.
    if is_frozen:
        import pyi_splash

    # Modify text
    for i in reversed(range(5)):
        print(f"Updating text (1st time) in {i+1} seconds...")
        sleep(1)
    if is_frozen:
        pyi_splash.update_text("This is a test text")

    # Modify text again
    for i in reversed(range(5)):
        print(f"Updating text (2nd time) in {i+1} seconds...")
        sleep(1)
    if is_frozen:
        pyi_splash.update_text("Second time's a charm")

    # Close the splash screen
    for i in reversed(range(5)):
        print(f"Closing splash screen in {i+1} seconds...")
        sleep(1)
    if is_frozen:
        pyi_splash.close()

    # Exit application
    for i in reversed(range(10)):
        print(f"Exiting application in {i+1} seconds...")
        sleep(1)


if __name__ == '__main__':
    main()
```

**Comment by bwoodsend** [comment, marked as answer]
In all honesty I think that all you'd really want is:
```python
# Import all your biggest dependencies here so that the splash screen is still visible whilst they're loading. e.g.
# import PyQt5
# import numpy

# If we're running in a frozen onefile application:
try:
    # Close the splash screen.
    import pyi_splash
    pyi_splash.close()
except ImportError:
    # Otherwise do nothing.
    pass

# Rest of your code here...
print("hello")
```
Which you'd build with:
```
pyinstaller --onefile --splash=my-splash-image.png your-code.py
```

**Reply by aliencaocao** [reply]
Thanks for the help, I made it work finally. I think that the documentation page needs a quick fix, because if you read it, it is very confusing to beginners as to what and where to wrap the things inside try except. The linked details page also provides no information on this.
![image](https://user-images.githubusercontent.com/20109683/131410495-db4f43e1-7b42-4a84-b2a7-16f23eaebf73.png)
